pallets/mixer/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+	type Group: GroupTrait<Self::AccountId, Self::BlockNumber, Self::GroupId>;
+	/// The max depth of the mixers
+	type MaxTreeDepth: Get<u8>;
+	/// The small deposit length
+	type DepositLength: Get<Self::BlockNumber>;
+}
+
+
+#[derive(Encode, Decode, PartialEq)]
+pub struct MixerInfo<T: Config> {
+	pub minimum_deposit_length_for_reward: T::BlockNumber,
+	pub fixed_deposit_size: BalanceOf<T>,
+	pub leaves: Vec<Data>,
+}
+
+impl<T: Config> core::default::Default for MixerInfo<T> {
+	fn default() -> Self {
+		Self {
+			minimum_deposit_length_for_reward: Zero::zero(),
+			fixed_deposit_size: Zero::zero(),
+			leaves: Vec::new(),
+		}
+	}
+} 
+
+
+impl<T: Config> MixerInfo<T> {
+	pub fn new(min_dep_length: T::BlockNumber, dep_size: BalanceOf<T>, leaves: Vec<Data>) -> Self {
+		Self {
+			minimum_deposit_length_for_reward: min_dep_length,
+			fixed_deposit_size: dep_size,
+			leaves: leaves,
+		}
+	}
+}
+
+// This pallet's storage items.
+decl_storage! {
+	trait Store for Module<T: Config> as Mixer {
+		pub Initialised get(fn initialised): bool;
+		/// The map of mixer groups to their metadata
+		pub MixerGroups get(fn mixer_groups): map hasher(blake2_128_concat) T::GroupId => MixerInfo<T>;
+		/// Map of used nullifiers (Data) for each tree.
+		pub UsedNullifiers get(fn used_nullifiers): map hasher(blake2_128_concat) (T::GroupId, Data) => bool;
+	}
+}
+
+// The pallet's events
+decl_event!(
+	pub enum Event<T>
+	where
+		AccountId = <T as frame_system::Config>::AccountId,
+		GroupId = <T as merkle::Config>::GroupId,
+		Nullifier = Data,
+	{
+		Deposit(GroupId, AccountId, Nullifier),
+		Withdraw(GroupId, AccountId, Nullifier),
+	}
+);
+
+// The pallet's errors
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		/// Value was None
+		NoneValue,
+		///
+		NoMixerForId,
+		///
+		NotInitialised,
+		///
+		AlreadyInitialised,
+		///
+		AlreadyUsedNullifier,
+		///
+		MixerDoesntExist,
+		///
+		InsufficientBalance,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		#[weight = 0]
+		pub fn deposit(origin, mixer_id: T::GroupId, data_points: Vec<Data>) -> dispatch::DispatchResult {
+			let sender = ensure_signed(origin)?;
+			ensure!(Self::initialised(), Error::<T>::NotInitialised);
+			// get mixer info, should always exist if module is initialised
+			let mut mixer_info = Self::get_mixer(mixer_id)?;
+			// ensure the sender has enough balance to cover deposit
+			let balance = T::Currency::free_balance(&sender);
+			// TODO: Multiplication by usize should be possible
+			// using this hack for now, though we should optimise with regular multiplication
+			// `data_points.len() * mixer_info.fixed_deposit_size`
+			let deposit: BalanceOf<T> = data_points.iter()
+				.map(|_| mixer_info.fixed_deposit_size)
+				.fold(Zero::zero(), |acc, elt| acc + elt);
+			ensure!(balance >= deposit, Error::<T>::InsufficientBalance);
```

Comment:
These kinds of checks should be done inside of `Currency` implementation

Reply:
Sure, I left it in for now. But we can consider removing it in the future, probably when we add more tests.